src/compile/repeat.ts

```diff
@@ -0,0 +1,204 @@
+import {Channel} from '../channel';
+import {CellConfig, Config} from '../config';
+import {FieldDef} from '../fielddef';
+import {FILL_STROKE_CONFIG} from '../mark';
+import {Repeat} from '../repeat';
+import {RepeatSpec} from '../spec';
+import {Dict, duplicate, flatten, vals} from '../util';
+import {VgData, VgEncodeEntry, VgLayout, VgScale} from '../vega.schema';
+import {applyConfig, buildModel} from './common';
+import {assembleData} from './data/assemble';
+import {assembleLayoutData} from './layout';
+import {Model} from './model';
+
+
+export type RepeatValues = {
+  row: string,
+  column: string
+};
+
+export class RepeatModel extends Model {
+  public readonly repeat: Repeat;
+
+  public readonly children: Model[];
+
+  /**
+   * Fixed width for the unit visualization.
+   * If undefined (e.g., for ordinal scale), the width of the
+   * visualization will be calculated dynamically.
+   */
+  public readonly width: number;
+
+  /**
+   * Fixed height for the unit visualization.
+   * If undefined (e.g., for ordinal scale), the height of the
+   * visualization will be calculated dynamically.
+   */
+  public readonly height: number;
+
+  constructor(spec: RepeatSpec, parent: Model, parentGivenName: string, repeatValues: RepeatValues, config: Config) {
+    super(spec, parent, parentGivenName, config);
+
+    this.repeat = spec.repeat;
+    this.children = this._initChildren(spec, this.repeat, repeatValues, config);
+  }
+
+  private _initChildren(spec: RepeatSpec, repeat: Repeat, repeatValues: RepeatValues, config: Config): Model[] {
+    const children: Model[] = [];
+    const row = repeat.row || [repeatValues ? repeatValues.row : null];
+    const column = repeat.column || [repeatValues ? repeatValues.column : null];
+
+    // cross product
+    for (const r in row) {
+      if (row.hasOwnProperty(r)) {
+        const rowField = row[r];
+        for (const c in column) {
+          if (column.hasOwnProperty(c)) {
+            const columnField = column[c];
+
+            const name = (rowField ? '_' + rowField : '') + (columnField ? '_' + columnField : '');
+
+            const childRepRepeatValues = {
+              row: rowField,
+              column: columnField
+            };
+
+            children.push(buildModel(duplicate(spec.spec), this, this.getName('child' + name), childRepRepeatValues, config));
```

Comment:
Do we really need to duplicate?
It looks like that your `encoding/facetRepeatResolve` do not cause side effect anyway.